server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs

```diff
@@ -48,43 +52,69 @@ import           Hasura.RQL.Types
 import           Hasura.SQL.Value
 
 -- | https://graphql.github.io/graphql-spec/June2018/#sec-Response-Format
+--
+-- NOTE: this type and parseGQRespValue are a lax representation of the spec,
+-- since...
+--   - remote GraphQL servers may not conform strictly, and...
+--   - we use this type as an accumulator.
+--
+-- Ideally we'd have something correct by construction for hasura results
+-- someplace.
 data GQRespValue =
   GQRespValue
-  { gqRespData   :: Maybe OJ.Object
-  , gqRespErrors :: Maybe [OJ.Value]
-  } deriving (Show, Eq)
-
-parseGQRespValue :: OJ.Value -> Either String GQRespValue
-parseGQRespValue =
-  \case
-    OJ.Object obj -> do
-      gqRespData <-
-        case OJ.lookup "data" obj of
-          Nothing -> pure Nothing
-          Just (OJ.Object dobj) -> pure (Just dobj)
-          Just _ -> Left "expected object for GraphQL data response"
-      gqRespErrors <-
-        case OJ.lookup "errors" obj of
-          Nothing -> pure Nothing
-          Just (OJ.Array vec) -> pure (Just (toList vec))
-          Just _ -> Left "expected array for GraphQL error response"
-      pure (GQRespValue {gqRespData, gqRespErrors})
-    _ -> Left "expected object for GraphQL response"
-
-gqrespValueToValue :: GQRespValue -> OJ.Value
-gqrespValueToValue (GQRespValue mdata' merrors) =
-  OJ.Object
-    (OJ.fromList
-       (concat
-          [ [("data", OJ.Object data') | Just data' <- [mdata']]
-          , [ ("errors", OJ.Array (V.fromList errors))
-            | Just errors <- [merrors]
-            ]
-          ]))
-
-data GQJoinError =  GQJoinError !T.Text
-  deriving (Show, Eq)
-
+  { _gqRespData   :: OJ.Object
+  -- ^ 'OJ.empty' (corresponding to the invalid `"data": {}`) indicates an error.
+  , _gqRespErrors :: VB.Builder OJ.Value
+  -- ^ An 'OJ.Array', but with efficient cons and concatenation. Null indicates
+  -- query success.
+  }
+-- TODO consider unifying this type with Hasura.GraphQL.Transport.HTTP.Protocol.
+
+makeLenses ''GQRespValue
+
+parseGQRespValue :: EncJSON -> Either String GQRespValue
+parseGQRespValue = OJ.eitherDecode . encJToLBS >=> \case
+  OJ.Object obj -> do
+    _gqRespData <-
+      case OJ.lookup "data" obj of
+        -- "an error was encountered before execution began":
+        Nothing -> pure OJ.empty
```

Comment:
should we not return null here instead of `{}` ?

Reply:
So after these changes in the context of `GQRespValue` we use `OJ.empty` to represent all the error cases with `data`. The idea is that prior to this change the type could represent the error/missing case in several different ways, which we had to handle in various places, e.g. there was no meaningful difference between `Nothing` and `Just OJ.empty`.

I hope that makes sense, I've been on a red-eye flight ...

Reply:
That makes sense. Rest your eyes :)